x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/WeightedTokensQueryBuilder.java

```diff
@@ -0,0 +1,263 @@
+        out.writeCollection(tokens);
+        out.writeOptionalWriteable(tokenPruningConfig);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.startObject(fieldName);
+        builder.startObject(TOKENS_FIELD.getPreferredName());
+        for (var token : tokens) {
+            token.toXContent(builder, params);
+        }
+        builder.endObject();
+        if (tokenPruningConfig != null) {
+            tokenPruningConfig.toXContent(builder, params);
+        }
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+        builder.endObject();
+    }
+
+    /**
+     * We calculate the maximum number of unique tokens for any shard of data. The maximum is used to compute
+     * average token frequency since we don't have a unique inter-segment token count.
+     * Once we have the maximum number of unique tokens, we use the total count of tokens in the index to calculate
+     * the average frequency ratio.
+     *
+     * @param reader
+     * @param fieldDocCount
+     * @return float
+     * @throws IOException
+     */
+    private float getAverageTokenFreqRatio(IndexReader reader, int fieldDocCount) throws IOException {
+        int numUniqueTokens = 0;
+        for (var leaf : reader.getContext().leaves()) {
+            var terms = leaf.reader().terms(fieldName);
+            if (terms != null) {
+                numUniqueTokens = (int) Math.max(terms.size(), numUniqueTokens);
+            }
+        }
+        if (numUniqueTokens == 0) {
+            return 0;
+        }
+        return (float) reader.getSumDocFreq(fieldName) / fieldDocCount / numUniqueTokens;
+    }
+
+    /**
+     * Returns true if the token should be queried based on the {@code tokensFreqRatioThreshold} and {@code tokensWeightThreshold}
+     * set on the query.
+     */
+    private boolean shouldKeepToken(
+        IndexReader reader,
+        WeightedToken token,
+        int fieldDocCount,
+        float averageTokenFreqRatio,
+        float bestWeight
+    ) throws IOException {
+        if (this.tokenPruningConfig == null) {
+            return true;
+        }
+        int docFreq = reader.docFreq(new Term(fieldName, token.token()));
+        if (docFreq == 0) {
+            return false;
+        }
+        float tokenFreqRatio = (float) docFreq / fieldDocCount;
+        return tokenFreqRatio < this.tokenPruningConfig.getTokensFreqRatioThreshold() * averageTokenFreqRatio
+            || token.weight() > this.tokenPruningConfig.getTokensWeightThreshold() * bestWeight;
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        final MappedFieldType ft = context.getFieldType(fieldName);
+        if (ft == null) {
+            return new MatchNoDocsQuery("The \"" + getName() + "\" query is against a field that does not exist");
+        }
```

Comment:
Shouldn't we check the field type? It seems like this should only ever be executed against rank_features or sparse_vector.

Reply:
@kderusso if expanded_tokens doesn't make this check, let's skip it for now. No need to bog this down.

But, it seems to me we should check it for both :).